lib/ansible/modules/cloud/amazon/aws_s3_bucket_facts.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Ansible Project
+# This is a free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This Ansible library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: aws_s3_bucket_facts
+short_description: Lists S3 buckets in AWS
+requirements:
+  - boto3 >= 1.4.4
+  - python >= 2.6
+description:
+    - Lists S3 buckets in AWS
+version_added: "2.4"
+author: "Gerben Geijteman (@hyperized)"
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+
+# Note: Only AWS S3 is currently supported
+
+# Lists all s3 buckets
+- aws_s3_bucket_facts:
+'''
+
+RETURN = '''
+Buckets:
+  description: List of buckets
+  returned: always
+  type: list
+  sample:
+    [{
+        "Name": "my_bucket",
+        "CreationDate": "2017-07-06T15:05:12+00:00"
+    }]
+Owner:
+  description: Details of bucket owner
+  returned: always
+  type: dict
+  sample:
+    {
+      "DisplayName": "my_user",
+      "ID": "d5a0152fc8eb469fcc7eed35b2da939fd9e1a9c53924b681450f5353199900f8"
+    }
+ResponseMetadata:
+  description: Details of API call
+  returned: always
+  type: dict
+  sample:
+    {
+      "HTTPHeaders": {
+        "content-type": "application/xml",
+        "date": "Thu, 27 Jul 2017 11:37:07 GMT",
+        "server": "AmazonS3",
+        "transfer-encoding": "chunked",
+        "x-amz-id-2": "LeNLtFwr0DmQD7+pb52w3A+vuR9vkVNDswfJyTHj6673yBDoYS24EShCztue+ARrxqzDGsNWmOk=",
+        "x-amz-request-id": "F47C7B7622231E44"
+      },
+      "HTTPStatusCode": 200,
+      "HostId": "LeNLtFwr0DmQD7+pb52w3A+vuR9vkVNDswfJyTHj6673yBDoYS24EShCztue+ARrxqzDGsNWmOk=",
+      "RequestId": "F47C7B7622231E44",
+      "RetryAttempts": 0
+    }
+'''
+
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, HAS_BOTO3, camel_dict_to_snake_dict, \
+    get_aws_connection_info
+
+try:
+    import botocore
+except ImportError:
+    pass  # will be detected by imported HAS_BOTO3
+
+
+class S3ServiceManager(object):
```

Comment:
Classes don't inherently make things easier to extend, since in a module you'rd basically storing state in `self.[name]` in a program where there's only ever one instance.